File: tools/claude/config/skills/scanning-git-for-tils/git/formatting.ts
/**
 * Git commit formatting utilities.
 */

import type { Commit } from "./commits.ts";

export function shouldSkipCommit(commit: Commit): boolean {
  const subject = commit.subject.toLowerCase();

  // Skip dependabot commits
  if (subject.includes("bump") && subject.includes("from")) return true;

  // Skip merge commits
  if (subject.startsWith("merge")) return true;

  return false;
}

export function formatMarkdown(
  commits: Commit[],
  days: number,
  newCount: number,
  totalCount: number,
): string {
  const lines: string[] = [];

  // Header
  if (totalCount === 0) {
    lines.push(`# Git commits from last ${days} days:\n`);
    lines.push("No commits found.\n");
    return lines.join("\n");
  }

  if (newCount === 0) {
    lines.push(`# Git commits from last ${days} days:\n`);
    lines.push(
      `No new commits to assess (${totalCount} commits already reviewed).\n`,
    );
    return lines.join("\n");
  }

  const reviewedCount = totalCount - newCount;
  const statusSuffix = reviewedCount > 0
    ? ` (${newCount} new, ${reviewedCount} already reviewed)`
    : "";
  lines.push(`# Git commits from last ${days} days:${statusSuffix}\n`);

  // Format each commit
  commits.forEach((commit, index) => {
    lines.push(`${index + 1}. [${commit.repo}] ${commit.subject}`);
    lines.push(`   Hash: ${commit.hash} (index: ${index}) | Date: ${commit.date}`);

    if (commit.body) {
      const truncated = commit.body.length > 200 ? commit.body.slice(0, 200) + "..." : commit.body;
      lines.push(`   Body: ${truncated}`);
    }

    const fileDisplay = commit.files.length === 0
      ? "(no files)"
      : commit.files.length > 5
      ? `${commit.files.slice(0, 5).join(", ")} (+${commit.files.length - 5} more)`
      : commit.files.join(", ");
    lines.push(`   Files: ${fileDisplay}`);
    lines.push(`   URL: ${commit.url}\n`);
  });

  return lines.join("\n");
}

File: tools/claude/config/skills/scan-git-for-tils/git/formatting.py
"""Git commit formatting utilities."""

from __future__ import annotations

from datetime import datetime

from git.types import Commit


def format_relative_date(iso_date: str) -> str:
    """Convert ISO date to relative format."""
    if not iso_date:
        return "unknown"

    try:
        dt = datetime.fromisoformat(iso_date.replace("Z", "+00:00"))
        now = datetime.now(dt.tzinfo)
        diff = now - dt

        if diff.days == 0:
            hours = diff.seconds // 3600
            if hours == 0:
                return "just now"
            return f"{hours} hour{'s' if hours != 1 else ''} ago"
        elif diff.days == 1:
            return "yesterday"
        elif diff.days < 7:
            return f"{diff.days} days ago"
        elif diff.days < 30:
            weeks = diff.days // 7
            return f"{weeks} week{'s' if weeks != 1 else ''} ago"
        else:
            months = diff.days // 30
            return f"{months} month{'s' if months != 1 else ''} ago"
    except (ValueError, TypeError):
        return "unknown"


def should_skip_commit(commit: Commit) -> bool:
    """Check if commit should be filtered out entirely."""
    subject = commit.subject.lower()

    # Skip dependency bot commits
    if "dependabot" in subject or ("bump" in subject and "from" in subject):
        return True

    # Skip merge commits
    if subject.startswith("merge"):
        return True

    return False


def format_markdown(commits: list[Commit], days: int, new_count: int, total_count: int) -> str:
    """Format commits as markdown for Claude to evaluate."""
    header = f"Git commits from last {days} days:\n"

    if total_count > 0 and new_count == 0:
        return f"{header}\nNo new commits to assess ({total_count} commits already reviewed)."

    if not commits:
        return f"{header}\nNo commits found. Try increasing the date range."

    lines = [header]
    if new_count < total_count:
        lines.append(f"({new_count} new, {total_count - new_count} already reviewed)\n")

    for i, commit in enumerate(commits, 1):
        files_str = ", ".join(commit.files[:5]) if commit.files else "(no files)"
        if len(commit.files) > 5:
            files_str += f" (+{len(commit.files) - 5} more)"

        lines.append(f"{i}. [{commit.repo}] {commit.subject}")
        lines.append(f"   Hash: {commit.hash} (index: {i-1}) | Date: {commit.date}")
        if commit.body:
            body_preview = commit.body[:200] + "..." if len(commit.body) > 200 else commit.body
            lines.append(f"   Body: {body_preview}")
        lines.append(f"   Files: {files_str}")
        lines.append(f"   URL: {commit.url}")
        lines.append("")

    return "\n".join(lines)
